File: android/app/src/main/java/com/emojicode/app/cipher/CipherUrlCodec.kt
package com.emojicode.app.cipher

import android.util.Base64
import org.json.JSONObject
import android.net.Uri

/**
 * Web版と同じURLフォーマット:
 *   https://*.github.io/mencon-emoji-henkan/#cipher=BASE64URL(JSON({v:1, c:差分, x:特殊}))
 *
 * カスタムスキームでも同様:
 *   emojicode://import?cipher=BASE64URL(...)
 */
object CipherUrlCodec {
    private const val VERSION = 1

    /** URI / 任意のテキストから暗号表(差分込み)を抽出。なければ null */
    fun parseFromUri(uri: Uri?): Map<String, String>? {
        if (uri == null) return null
        val raw = uri.toString()
        val regex = Regex("[#?&]cipher=([^&\\s]+)")
        val m = regex.find(raw) ?: return null
        return parsePayload(m.groupValues[1])
    }

    fun parseFromText(text: String?): Map<String, String>? {
        if (text.isNullOrBlank()) return null
        val regex = Regex("[#?&]cipher=([^&\\s]+)")
        val m = regex.find(text) ?: return null
        return parsePayload(m.groupValues[1])
    }

    private fun parsePayload(encoded: String): Map<String, String>? {
        return try {
            val padded = encoded.padEnd(encoded.length + (4 - encoded.length % 4) % 4, '=')
            val bytes = Base64.decode(padded, Base64.URL_SAFE)
            val json = JSONObject(String(bytes, Charsets.UTF_8))
            if (json.optInt("v") != VERSION) return null
            val result = HashMap(DefaultCipher.MAP)
            json.optJSONObject("c")?.let { diff ->
                diff.keys().forEach { k -> result[k] = diff.getString(k) }
            }
            json.optJSONObject("x")?.let { extra ->
                extra.keys().forEach { k -> result[k] = extra.getString(k) }
            }
            result
        } catch (e: Exception) { null }
    }

    fun buildShareUrl(cipher: Map<String, String>, baseUrl: String): String {
        val diff = JSONObject()
        val extra = JSONObject()
        for ((k, v) in cipher) {
            val def = DefaultCipher.MAP[k]
            if (def != null) {
                if (def != v) diff.put(k, v)
            } else extra.put(k, v)
        }
        val payload = JSONObject().apply {
            put("v", VERSION); put("c", diff); put("x", extra)
        }.toString()
        val encoded = Base64.encodeToString(
            payload.toByteArray(Charsets.UTF_8),
            Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP
        )
        return "$baseUrl#cipher=$encoded"
    }
}

File: ios/EmojiCode/Cipher/CipherUrlCodec.swift
import Foundation

/// Web版と完全互換のURL/QR共有フォーマット
public enum CipherUrlCodec {
    private static let version = 1

    /// URI文字列から暗号表を抽出
    public static func parse(from urlString: String) -> [String: String]? {
        let pattern = "[#?&]cipher=([^&\\s]+)"
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: urlString,
                                            range: NSRange(urlString.startIndex..., in: urlString)),
              let range = Range(match.range(at: 1), in: urlString) else {
            return nil
        }
        let encoded = String(urlString[range])
        return parsePayload(encoded)
    }

    private static func parsePayload(_ encoded: String) -> [String: String]? {
        // base64url → base64
        var b64 = encoded.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let mod = b64.count % 4
        if mod != 0 { b64 += String(repeating: "=", count: 4 - mod) }

        guard let data = Data(base64Encoded: b64),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let v = json["v"] as? Int, v == version else {
            return nil
        }
        var result = DefaultCipher.map
        if let diff = json["c"] as? [String: String] {
            for (k, v) in diff { result[k] = v }
        }
        if let extra = json["x"] as? [String: String] {
            for (k, v) in extra { result[k] = v }
        }
        return result
    }

    /// 暗号表をシェア用URLに変換
    public static func buildShareUrl(cipher: [String: String],
                                      baseUrl: String = "https://yu08083.github.io/mencon-emoji-henkan/") -> String {
        var diff = [String: String]()
        var extra = [String: String]()
        for (k, v) in cipher {
            if let def = DefaultCipher.map[k] {
                if def != v { diff[k] = v }
            } else {
                extra[k] = v
            }
        }
        let payload: [String: Any] = ["v": version, "c": diff, "x": extra]
        guard let data = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]),
              let encoded = data.base64URLEncodedString() else {
            return baseUrl
        }
        return "\(baseUrl)#cipher=\(encoded)"
    }
}

private extension Data {
    func base64URLEncodedString() -> String? {
        return self.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
